### ISSUE TITLE
module initialize override

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#9335]
# Submitting questions

If you have a question like "how do I do X?", this is not the right place for asking it. Please ask on the [Crystal Forum](https://forum.crystal-lang.org), on our combined [Gitter](https://gitter.im/crystal-lang/crystal)/[IRC](http://webchat.freenode.net/?channels=#crystal-lang) or on [StackOverflow](http://stackoverflow.com/questions/tagged/crystal-lang).

# Submitting bugs

Make sure to review these points before submitting issues - thank you!

- Make sure a similar issue doesn't exist yet: use the search box
- **Include reproducible code**: we should be able to paste it into an editor, compile and run it and get the same error as you. Otherwise it's impossible for us to reproduce the bug.
- Don't **only** use `play.crystal-lang.org` or `carc.in`: code might be lost and the issue will remain incomplete. Write code in the issue itself.
- Reduce code, if possible, to the minimum size that reproduces the bug.
- If all of the above is impossible due to a large project, create a branch that reproduces the bug and point us to it.
- Include Crystal compiler version (`crystal -v`) and OS. If possible, try to see if the bug still reproduces on master.

**Comment by asterite** [comment]
It might be unexpected for you, but it's totally expected for me and I coded it that way. If Mixin has instance variables that must be initialized, bypassing its constructor will bypass the initialization of those instance variables.

Maybe put it as a feature request or discussion, because it's not a bug (it's not something I didn't think of).

**Comment by straight-shoota** [comment]
Hm, but it's at least a bug with `JSON::Serializable`, don't you think? That should work even if a parent class doesn't include it.

**Comment by asterite** [comment]
Sorry, I don't follow. Can you show an example with JSON::Serializable that doesn't work?

**Comment by asterite** [comment]
By the way, a way to make it work is to use macro included and inject the initialize in the including module. And that's what JSON::Serializable does.

**Comment by asterite** [comment]
Another way to explain the situation: initialize and new are never inherited if another type sooner in the hierarchy defines them.

**Comment by straight-shoota** [comment]
```cr
require "json"
 
class Parent
  def initialize(@foo : String)
  end
end
 
class Child < Parent
  include JSON::Serializable
end
 
Child.new("foo") # Error: no overload matches 'Child.new' with type String
```

**Comment by asterite** [comment]
Well, it defines new constructors and constructors are not inherited.

I already proposed doing deserialization not using constructors and it was rejected. Maybe it's a bit more clear now why I wanted that.

**Comment by asterite** [comment]
I'm this particular case, just define the constructor again and call super. Really easy to solve.

**Comment by straight-shoota** [comment]
I've tried that, but `super` does not work either because it seems to only resolve to `Serializable::initialize`:
```cr
require "json"
 
class Parent
  def initialize(@foo : String)
  end
end
 
class Child < Parent
  include JSON::Serializable
  def initialize(foo : String)
    super # Error: missing argument: __pull_for_json_serializable
  end
end
 
Child.new("foo")
```

**Comment by asterite** [comment]
super(foo) doesn't work either?

Maybe make the parent Serializable?

**Comment by straight-shoota** [comment]
`super(foo)` doesn't work either. And I'm pretty shure it shouldn't if `super` doesn't work.

Including `Serializable` in `Parent` could work in a way, but it also includes the module in other subtypes which are not supposed to be `Serializable`. So it's not a real solution.

**Comment by asterite** [comment]
> but it also includes the module in other subtypes which are not supposed to be Serializable

If you are not supposed to call `to_json` on them, just don't call it.

In fact, I still can't still the problem of having all types be serializable to JSON by default like in Java, C#, Go, etc., and where you would control **how** (and if) they are serialized.

Also, the above code in Ruby works exactly like in Crystal. You just can't call super to reach `Parent`.

The solution is to define `def initialize(foo)` again and initialize `@foo` in it. It's simple too.

**Comment by RX14** [comment]
`super(foo: foo)` will work.

**Comment by straight-shoota** [comment]
Nope, that's also `missing argument: __pull_for_json_serializable`.

**Comment by jhass** [comment]
I do feel like this is a bug. Not inheriting constructors by default seems fine, but calling them with `super` should always be possible IMO, no matter the amount of inheritance levels in-between.
